In [170]:
## load required libraries 
import sklearn
from sklearn.feature_extraction.text import TfidfVectorizer
from pdpbox import pdp, get_dataset, info_plots
from eli5.sklearn import PermutationImportance
from sklearn.preprocessing import LabelEncoder
from collections import Counter
import pandas as pd 
import numpy as np 
import eli5
import re



In [198]:
#Load the dataset
df = pd.read_csv("../input/ks-projects-201801.csv",  parse_dates = ["launched", "deadline"])
df_copy = df.copy()  #keep a copy of the data frame that remains intact
print ("Number of Campaigns: ", projects.shape[0], "\nNumber of Columns: ", projects.shape[1])
df.head()



Number of Campaigns:  378661 
Number of Columns:  15


ID      ...      usd_goal_real
0  1000002330      ...            1533.95
1  1000003930      ...           30000.00
2  1000004038      ...           45000.00
3  1000007540      ...            5000.00
4  1000011046      ...           19500.00

[5 rows x 15 columns]

As of every machine learning problem, we need to explore the data and look out for problematic cases. These steps can be listed below:
* Outliers
* Missing values
* Unwanted information/labels
* Encode categorical values to numerical
* Extract Features

Let's see if there are any missing values. We will deal with outliers by considering features relevant to mean and median of each category. 

In [199]:
df.isnull().sum().sort_values(ascending = False)

usd pledged         3797
name                   4
usd_goal_real          0
usd_pledged_real       0
country                0
backers                0
state                  0
pledged                0
launched               0
goal                   0
deadline               0
currency               0
main_category          0
category               0
ID                     0
dtype: int64

So as it shows, usd_pledged has around 1% of it's values missing and there are 4 campaigns with missing names. Let's take a further look.

In [200]:
display(df[df['usd pledged'].isnull()].shape)
df[df['usd pledged'].isnull()].head()

(3797, 15)

ID      ...      usd_goal_real
169  1000694855      ...            6500.00
328   100149523      ...            3211.53
632  1003023003      ...            3500.00
647  1003130892      ...            6000.00
749  1003629045      ...            3000.00

[5 rows x 15 columns]

It seems reasonable to keep these 3797 rows and substitute the NAN values with the corresponding usd_pledged_real and drop those 4 rows with no name. 
Also a glance at the country column shows that some country entries seem to be incorrect with the '*N,0"*' entry. We should also consider taking care of that. 
Since most of the campaigns are from the states (78.85%) we will use a indicator of being from the states or not. Hence these values will turn to 0 (not from the states, assuming that '*N,0"*' does not indicate that).

In [201]:
df['usd pledged'].fillna(df['usd_pledged_real'], inplace=True)
df.dropna(inplace=True)

In [202]:
#Let's make sure the null values have been taken care of:
df.isnull().sum().sort_values(ascending = False)

usd_goal_real       0
usd_pledged_real    0
usd pledged         0
country             0
backers             0
state               0
pledged             0
launched            0
goal                0
deadline            0
currency            0
main_category       0
category            0
name                0
ID                  0
dtype: int64

Great! So now that we have dealt with the missing values, let's delve into the feature extraction step. Below is a list of useful features that we will be using:
*Features focusing on the name of the campaign can be:*
* Number of words in the name
* Number of syllables
* Presence of particular punctuation (e.g. ? !)
* Number of characters
* Ratio between vowels and alphanumeric length of the name
* Whether the name is uppercase or not (Maybe names in uppercase can attract more attention to potential backers)

These features are useful because we are interested to know whether a specific range of words, characters or any sort of pattern is effective in the campaign's success.
Other important features could be those relevant to the time of launch and its durations. There may be a correlation between the time and day/month a campaign is launched.
Some features in this aspect could be:
* Launched hour
* Launched day
* Launched week
* Launched month
* Whether it was a weekend or not
* Duration of the campaign 

Another group of features which is obviously effective is based on the goal and pledged amount. We can extract new features by considering the mean goal in each category and main category to see if a specific campaign is an outlier compared to other campaignes of its field, etc. Below is a list of such features:
* Mean of the category's goal
* Median of the category's goal
* Count of the categoty (to see whether the campaign is relatively popular or not)
* Mean of the main category's goal
* Median of the main category's goal
* Count of the main category
* Difference between a campaign's goal and mean of its category
* Pledged amount as a percentage of the goal
* Pledged amount per backer 
* Average Pledged amount per category 
* Difference between pledged amount per backer for a campaign and the average received in that category

Another interesting feature could be the country where the campaign is launched.

The following blocks computes the features listed above.

In [203]:
 #Let's only keep the successful and failed labels and turn it into a binary classification:
df.query("state in ['failed', 'successful']", inplace=True)

In [204]:
def syllable_count(word):
    """Count the syllables in a word"""
    word = word.lower()
    vowels = "aeiouy"
    count = 0
    if word[0] in vowels:
        count += 1
    for index in range(1, len(word)):
        if word[index] in vowels and word[index - 1] not in vowels:
            count += 1
    if word.endswith("e"):
        count -= 1
    if count == 0:
        count += 1
    return count

In [205]:
def countVowels2LettersRatio(word):
    '''Count ratio between vowels and letters'''
    word = str(word)
    count = 1  
    vowels = 0
    for i in word:
        if i.isalpha():
            count = count + 1
            if i in 'aeiou':
                vowels = vowels + 1
    return ((vowels * 1.0) / count)



In [206]:
## feature extraction for the campaign name

df["num_syllable"] = df["name"].apply(lambda x: syllable_count(x))
df["num_words"] = df["name"].apply(lambda x: len(x.split()))
df["num_chars"] = df["name"].apply(lambda x: len(x.replace(" ","")))
df['has_exclaimation_mark'] = int(len(re.findall(r'\!', str(df.name.str[:]))) != 0)  #presence of ! in the name
df['has_question_mark'] = int(len(re.findall(r'\?', str(df.name.str[:]))) != 0)  #presence of ? in the name
df['name_is_upper'] = df.name.str.isupper().astype(float)# if name is uppercase
df['name_vowel_ratio'] = df.name.apply(countVowels2LettersRatio)# for each name calculate vowels ratio




## feature extraction for the campaign length and timing
df["launched_month"] = df["launched"].dt.month
df["launched_week"] = df["launched"].dt.week
df["launched_day"] = df["launched"].dt.weekday
df["is_weekend"] = df["launched_day"].apply(lambda x: 1 if x > 4 else 0)
df["duration_in_days"] = df["deadline"] - df["launched"]
df["duration_in_days"] = df["duration_in_days"].apply(lambda x: int(str(x).split()[0]))

## label encoding the categorical features. This is necessary for the features related to goal and pledged amounts
#df = pd.concat([df, pd.get_dummies(df["main_category"])], axis = 1)
label_encoder = LabelEncoder()
for cat in ["category", "main_category"]:
    df[cat] = label_encoder.fit_transform(df[cat])
   
## feature extraction for goal and pledged values
df.head()
c1 = df.groupby("main_category").agg({"goal" : "mean", "category" : "sum"})
c2 = df.groupby("category").agg({"goal" : "mean", "main_category" : "sum"})
c1 = c1.reset_index().rename(columns={"goal" : "main_category_goal_mean", "category" : "main_category_count"})
c2 = c2.reset_index().rename(columns={"goal" : "category_goal_mean", "main_category" : "category_count"})
df = df.merge(c2, on = "category")
df = df.merge(c1, on = "main_category") 
df["diff_mean_category_goal"] = df["category_goal_mean"] - df["goal"]
df["diff_mean_category_goal"] = df["main_category_goal_mean"] - df["goal"]
df['goal_log'] = np.log1p(df.goal) # normalizing goal by applying log to prevent skewness
df.loc[:,'goal_reached'] = df['pledged'] / df['goal'] # Pledged amount as a percentage of goal.
df.loc[df['backers'] == 0, 'backers'] = 1 # In backers column, impute 0 with 1 to prevent undefined division.
df.loc[:,'pledge_per_backer'] = df['pledged'] / df['backers'] # Pledged amount per backer.

## create indicator variable for `country` variable
df['country_is_us'] = (df.country == 'US').astype(int)


In [207]:
df["state"] = df["state"].apply(lambda x: 1 if x=="successful" else 0)


In [208]:

df = df.drop(["launched", "deadline", "name", "currency", "country"], axis = 1) #drop the unnecassary columns

In [214]:
#Define the label and features and scale the data
label = df.state
features = [c for c in df.columns if c not in ["state"]]
#Let's normalise the data entries
df_scaled = pd.DataFrame(sklearn.preprocessing.normalize(df[features]))
#df_scaled.columns=list(df_scaled[features])

In [216]:
df_scaled.head()

0             1       ...                 28            29
0  0.999997  1.079994e-07      ...       0.000000e+00  0.000000e+00
1  0.999997  1.078888e-07      ...       1.997940e-08  0.000000e+00
2  0.999997  1.078414e-07      ...       5.408710e-08  9.985311e-10
3  0.999997  1.077850e-07      ...       5.039948e-08  9.980095e-10
4  0.999997  1.077283e-07      ...       1.828721e-08  0.000000e+00

[5 rows x 30 columns]

Now we are ready to split the data set to test and train and implement a model!

Resources: 
* https://www.kaggle.com/shivamb/an-insightful-story-of-crowdfunding-projects
* https://www.kaggle.com/srishti280992/kickstarter-project-classification-lgbm-70-3
* https://www.kaggle.com/dronqo/how-successful-is-your-kickstarter-project
* https://www.kaggle.com/kromel/kickstarter-successful-vs-failed
* https://www.kaggle.com/carlolepelaars/exploration-of-kickstarter-data-2010-2017